### PR TITLE
Features/smhi root

### PR DESCRIPTION
**Pull request by docNord**
I figure it is time to merge this into main now, to keep the pull request reasonably sized.

Quite a lot of functionality added:

* Finding closest station to given latitude and longitude
* Finding stations within given radius of given latitude and longitude
* Finding stations in vicinity of city, either closest or within given radius

On top of that, umbrella function for getting data has been added. Also, data format from Metobs has been changed to datetime to accomodate using diff() when building interpolation strategies.

I figure it is best to merge this back into main before developing further. Next step is to look at interpolation strategies and include mesan and strång functionality. Also, both unit and integration tests need to be added / updated.

Example to show one means by which I plan to include interpolation: Looking at nearby stations. When evaluating the snow depth around Kiruna, data is available starting from the late 19th century. Some stations lack data for extended periods in time, however. We can address the problem by looking at all stations around the city by running something like this code snippet:

```
import matplotlib.pyplot as plt
from smhi.smhi import SMHI

client = SMHI()
client.get_stations(3) # Causes error if not run. BUG!
client.find_stations_by_city(8,'Kiruna',20)

for nearby_station in client.nearby_stations:
    header, data = client.get_data(8,nearby_station[0])
    plt.plot(data['datetime'],data['Snödjup'])
plt.show()
```
snödjup_runt_kiruna

### CODE REVIEW

File: src/smhi/metobs.py [hunk 1]
```diff
@@ -126,6 +126,10 @@ def get_data(self, period: str = "corrected-archive") -> tuple[Any, Any]:
             on_bad_lines="skip",
             usecols=[0, 1, 2],
         )
+        table["datetime"] = table["Datum"] + ' ' + table["Tid (UTC)"]
```
Comment (mgcth) [inline src/smhi/metobs.py:129]: Looks good, but I guess both integration and unit tests need to be updated.

File: src/smhi/smhi.py [hunk 2]
```diff
@@ -78,11 +79,50 @@ def find_stations_from_gps(
...
-        all_stations = self.client.stations.data
-        self.d = [
-            s
-            for s in all_stations
-            if distance.distance(user_position, (s["latitude"], s["longitude"])) <= dist
-        ]
+        self.nearby_stations = []
+        all_stations = self.client.stations.stations
+        if dist == 0:
+            stations = [
```
Comment (mgcth) [inline src/smhi/smhi.py:85]: I would consider simplifying this or rewriting it as a normal loop. A bit hard to follow complex (multi-line) list comprehensions.

File: src/smhi/smhi.py [hunk 3]
```diff
@@ -78,11 +79,50 @@ def find_stations_from_gps(
...
+                (s['id'], s['name'], distance.distance(
+                    user_position, (s["latitude"], s["longitude"])).km)
+                for s in all_stations
+            ]
+            self.nearby_stations = min(stations, key=lambda x: x[2])
+        else:
+            self.nearby_stations = [
```
Comment (mgcth) [inline src/smhi/smhi.py:92]: Same here.

I think updating the unit tests will show how this method could be simplified.

File: src/smhi/smhi.py [hunk 4]
```diff
@@ -78,11 +79,50 @@ def find_stations_from_gps(
...
+
+    def find_stations_by_city(
+        self, parameter: int, city: str, dist: float = 0
+    ) -> None:
+        """Find stations for parameter from city name.
+
+        Args:
+            parameter: station parameter
+            dist: distance from city
+            city: name of city
```
Comment (mgcth) [inline src/smhi/smhi.py:108]: Same order as method parameters.

File: src/smhi/smhi.py [hunk 5]
```diff
@@ -78,11 +79,50 @@ def find_stations_from_gps(
...
+        """
+
```
Comment (mgcth) [inline src/smhi/smhi.py:110]: Linting error (no new line here).

File: src/smhi/smhi.py [hunk 6]
```diff
@@ -78,11 +79,50 @@ def find_stations_from_gps(
...
+        geolocator = Nominatim(user_agent='ifk-smhi')
+        loc = geolocator.geocode(city)
+        self.find_stations_from_gps(parameter=parameter, dist=dist,
+                                    latitude=loc.latitude, longitude=loc.longitude)
+
+    def get_data(
+        self, parameter: int, station: int, period: str = 'corrected-archive'
+    ) -> None:
```
Comment (mgcth) [inline src/smhi/smhi.py:118]: This method returns a tuple of something (is it str and some sort of list)? Strange that mypy doesn't fail here.

**Review by mgcth — CHANGES_REQUESTED** [review]
Nice additions! I think we should address unit and integration tests here, along with docs showing how to use these newly added features.

**Comment by docNord** [comment]
@mgcth could you help me resolve this?

**Comment by mgcth** [comment]
Is functionality done but missing tests?

**Comment by docNord** [comment]
I will fix it after work today. Functionality was done but I started over
to fix black and lint and tests, require some hours. You will have a PR to
review tonight 😊

**Comment by docNord** [comment]
Alright, the smhi feature works for Metobs. Should we incorporate fast access to the other subcategories as well?

Example:

```
from smhi.smhi import SMHI
s = SMHI()
s.find_stations_by_city(8,'Kiruna')
s.nearby_stations
h1,d1 = s.get_data(8,180960)
_ ,d2 = s.get_data(8,180960,interpolate=40)
``` 

Kiruna_snödjup

**Comment by docNord** [comment]
Metfcts would also benefit from functionality similar to the Metobs wrapper functions just developed, and I am sure Mesan and Strang as well - it will be very succint if you can ask our SMHI client for data from a named place and get it automatically.

How do we go about setting this up? @mgcth you have a parallell branch to align data output formats from the different modules, once that is in place it should be doable to align with minor adjustments.

**Comment by mgcth** [comment]
> Metfcts would also benefit from functionality similar to the Metobs wrapper functions just developed, and I am sure Mesan and Strang as well - it will be very succint if you can ask our SMHI client for data from a named place and get it automatically.
> 
> How do we go about setting this up? @mgcth you have a parallell branch to align data output formats from the different modules, once that is in place it should be doable to align with minor adjustments.

Create an issue about it, and we can talk more about design there.

File: src/smhi/smhi.py [hunk 7]
```diff
@@ -62,27 +64,127 @@ def get_stations_from_title(self, title: Optional[str] = None):
...
+
+    def find_stations_by_city(self, parameter: int, city: str, dist: float = 0) -> None:
+        """Find stations for parameter from city name.
+
+        Args:
+            parameter: station parameter
+            dist: distance from city
+            city: name of city
+        """
+
```
Comment (mgcth) [inline src/smhi/smhi.py:122]:
```suggestion

```
Fix linting.

**Comment by mgcth** [comment]
Haven't had the time to look at it. Will try this afternoon.

**Review by mgcth — APPROVED** [review]
If you agree with the exception, this can be merged.

**Comment by mgcth** [comment]
maybe you want to fix the `latin1` and `utf-8` encoding of the fixture file before merging. Also, do the docs need updating?

**Comment by docNord** [comment]
I don't know about the fixture file - why should we change it? The docs probably need editing, I could have a look quickly if you like.

**Comment by mgcth** [comment]
Because `latin1` encoding fails on your win machine... Let's fix it later.

**Comment by docNord** [comment]
Right - but why can't we just use utf8 as the encoding in the test? That works on my windows machine, is it causing problems on your side?

**Comment by docNord** [comment]
Also, looking at the docs: I'm thinking adding an example page for smhi would be good? Otherwise, I don't know where to add the stuff we've been working at here...

**Comment by mgcth** [comment]
@docNord, can you try running tests from the latest commit on your local computer?

**Comment by docNord** [comment]
The new variant works on my Windows machine too. Great work Mladen! With that, ok to merge? Will create a new PR, hopefully tomorrow, for adding place search to the other protocols.
